state-chain/pallets/cf-lending-pools/src/general_lending.rs

```diff
@@ -0,0 +1,1635 @@
+		extra_collateral: BTreeMap<Asset, AssetAmount>,
+	) -> Result<LoanId, DispatchError> {
+		ensure!(
+			T::SafeMode::get().borrowing_enabled.contains(&asset),
+			Error::<T>::LoanCreationDisabled
+		);
+
+		let config = LendingConfig::<T>::get();
+
+		let loan_id = NextLoanId::<T>::get();
+		NextLoanId::<T>::set(loan_id + 1);
+
+		LoanAccounts::<T>::mutate(&borrower_id, |maybe_account| {
+			let account = Self::create_loan_account_if_empty(
+				borrower_id.clone(),
+				maybe_account,
+				primary_collateral_asset,
+			)?;
+
+			let primary_collateral_asset = account.primary_collateral_asset;
+
+			ensure!(
+				extra_collateral.contains_key(&primary_collateral_asset),
+				Error::<T>::InvalidLoanParameters
+			);
+
+			account.primary_collateral_asset = primary_collateral_asset;
+
+			account.try_adding_collateral_from_free_balance(&borrower_id, &extra_collateral)?;
+
+			ensure!(
+				account.derive_ltv()? <= config.ltv_thresholds.target,
+				Error::<T>::InsufficientCollateral
+			);
+
+			GeneralLendingPools::<T>::try_mutate(asset, |pool| {
+				let pool = pool.as_mut().ok_or(Error::<T>::PoolDoesNotExist)?;
+
+				pool.provide_funds_for_loan(amount_to_borrow)?;
+
+				Ok::<_, DispatchError>(())
+			})?;
+
+			let mut loan = GeneralLoan {
+				asset,
+				created_at_block: frame_system::Pallet::<T>::current_block_number(),
+				owed_principal: amount_to_borrow,
+				fees_paid: BTreeMap::new(),
+			};
+
+			let origination_fee = Self::charge_origination_fee(
+				&borrower_id,
+				&mut loan,
+				primary_collateral_asset,
+				amount_to_borrow,
+			)?;
+
+			account.loans.insert(loan_id, loan);
+
+			T::Balance::credit_account(&borrower_id, asset, amount_to_borrow);
+
+			Self::deposit_event(Event::LoanCreated {
+				loan_id,
+				borrower_id: borrower_id.clone(),
+				asset,
+				principal_amount: amount_to_borrow,
+				origination_fee,
+			});
+
+			Ok::<_, DispatchError>(())
+		})?;
+
+		Ok(loan_id)
+	}
+
+	/// Borrows `extra_amount_to_borrow` by expanding `loan_id`. Adds any extra collateral to the
+	/// account (which may be required to cover the new total owed amount).
+	#[transactional]
+	fn expand_loan(
+		borrower_id: Self::AccountId,
+		loan_id: LoanId,
+		extra_amount_to_borrow: AssetAmount,
+		extra_collateral: BTreeMap<Asset, AssetAmount>,
+	) -> Result<(), DispatchError> {
+		let config = LendingConfig::<T>::get();
+
+		LoanAccounts::<T>::mutate(&borrower_id, |maybe_account| {
+			let loan_account = maybe_account.as_mut().ok_or(Error::<T>::LoanNotFound)?;
+
+			{
+				let loan = loan_account.loans.get_mut(&loan_id).ok_or(Error::<T>::LoanNotFound)?;
+
+				ensure!(
+					T::SafeMode::get().borrowing_enabled.contains(&loan.asset),
+					Error::<T>::LoanCreationDisabled
+				);
+
+				loan.owed_principal.saturating_accrue(extra_amount_to_borrow);
+			}
+
+			loan_account
+				.try_adding_collateral_from_free_balance(&borrower_id, &extra_collateral)?;
+
+			if loan_account.derive_ltv()? > config.ltv_thresholds.target {
```

Comment:
In new_loan we check the LTV *before* taking the loan, here we take it *after* increasing the owed principal. Is this inentional? What am I missing?